kernel/src/hil/nfc.rs

```diff
@@ -0,0 +1,105 @@
+    Off,
+    /// Both NFCT field events have been set - ambiguous state.
+    Unknown,
+}
+
+/// Controls an NFC tag main functionalities
+pub trait NfcTag<'a> {
+    /// Set the client instance that will handle callbacks
+    fn set_client(&self, client: &'a dyn Client<'a>);
+
+    /// Enable NFC sense field mode, and subscribe to the relevant interrupts.
+    /// Also set up the default configurations of how frame delay should be
+    /// dealt with (e.g. the maximum delay before a timeout).
+    /// This function should never fail.
+    fn enable(&self);
+
+    /// Enable tag emulation by triggering the necessary task.
+    /// This function should never fail.
+    fn activate(&self);
+    /// Notify the client by calling `field_lost()`. Then disable tag emulation
+    /// by triggering the necessary task. And go back to state of sense field.
+    /// This function should never fail.
+    fn deactivate(&self);
+
+    /// Pass the buffer to be transmitted and the amount of data and take
+    /// ownership of it. Subscribe to the relevant interrupt and trigger
+    /// the task for transmission.
+    ///     
+    /// On success returns the length of data to be sent.
+    /// On failure returns an error code and the buffer passed in.
+    fn transmit_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_amount: usize,
+    ) -> Result<usize, (ReturnCode, &'static mut [u8])>;
+
+    /// Pass a buffer for receiving data and take ownership of it.
+    ///     
+    /// On success returns nothing.
+    /// On failure returns an error code and the buffer passed in.
+    fn receive_buffer(
+        &self,
+        rx_buffer: &'static mut [u8],
+    ) -> Result<(), (ReturnCode, &'static mut [u8])>;
+
+    /// Configuration of the Tag according to its Type.
+    ///
+    /// Return `SUCCESS` if the tag type is supported and
+    /// `ENOSUPPORT` otherwise.
+    fn configure(&self, tag_type: u8) -> ReturnCode;
+
+    /// Set the maximum frame delay in number of 13.56 MHz clocks.
+    fn set_framedelaymax(&self, max_delay: u32);
+
+    /// Enable the interrupt for event SELECTED
+    fn unmask_select(&self);
+}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// Called when a selection event takes place.
+    /// This will call `set_framedelaymax()` to update
+    /// the default value in use.
+    fn tag_selected(&'a self);
+
+    // Reset the state automaton of the capsule.
+    fn tag_deactivated(&'a self);
+
+    /// Called when a field is detected.
+    /// This will notify the app of the presence of a field to activate the tag.
+    fn field_detected(&'a self);
+
+    /// Called when a field is lost.
+    /// This will notify the app of the absence of a field.
+    /// Returns any buffers passed either in `receive_buffer()` or `transmit_buffer()`
+    /// back to the capsule. Also trigger any ready callback functions.
```

Comment:
The buffer explanation here is great ✅ 

I'm not sure I quite understand which callbacks might be triggered here. Which ones are they, and will they be triggered in a way that indicates an error?